### ISSUE TITLE
[Proposal] Carousel Component

### ISSUE THREAD

**Issue by voronianski** [sumup-oss/circuit-ui#409]
Recently we've implemented a `Carousel` component on the marketing website.
 
It was made with re-usability in mind, so that in the future it can be ported to Circuit UI and other teams can benefit from reusing it as well.
 
## Visual

![carousel](https://user-images.githubusercontent.com/974035/59263835-20f26f00-8c42-11e9-9131-98159b244a73.gif)

You can also play with it on website's storybook page or see it in action as part of `Testimonials` component on [careers page](https://sumup.com/careers) for example.

## Context

Currently it's used only inside `Testimonials` component on the website but it will be also featured in a new fancy sliding `Hero` components as well.

Here's an example how it can be used on Dashboard:

![carousel on dashboard](https://user-images.githubusercontent.com/974035/59269883-52723700-8c50-11e9-9041-24587ae8d0cf.gif)

## Stateful vs. Composed versions

There's a _stateful_ version of component which just receives a bunch of props and handles all the animation and state management under the hood:
```js
import { Carousel } from '@sumup/circuit-ui';

const slidesData = [
  {  image: { src, alt } }, 
  {  image: { src, alt } }
  //...
];

<Carousel
  slides={slidesData}
  aspectRatio={1.8}
  slideDuration={5000}
  animationDuration={600}
  cycle
  swipe
  autoPlay
  pauseOnHover
/>
```

But carousel can also be used to _compose_ completely different slider using the provided components (though you will need to manage state by yourself):

```js
import { Carousel } from '@sumup/circuit-ui';

const { 
  Container, 
  Slides, 
  Slide, 
  SlideImage, 
  Controls, 
  Status,
  PrevButton,
  NextButton
} = Carousel

const slides = [
  {  image: { src, alt } }, 
  {  image: { src, alt } }
  //...
];

const CustomCarousel = () => {
  const total = slides.length;
  const [step, setStep] = useState(0);
  const goBack = () => setStep(step === 0 ? total - 1 : step - 1);
  const goForward = () => setStep(step === total - 1 ? 0 : step + 1);

  return (
    <Container>
      <Slides>
        {slides.map(({ image }, index) => (
          <Slide
            key={index}
            index={index}
            step={step}
          >
            <SlideImage
              src={image.src}
              alt={image.alt}
            />
          </Slide>
        ))}
      </Slides>
      <Controls>
        <PrevButton onClick={goBack} />
        <Status step={step} total={total}  />
        <NextButton onClick={goForward} />
      </Controls>
    </Container>
  );
};
```
...which will end up in something like -

![composed carousel](https://user-images.githubusercontent.com/974035/59271210-a0d50500-8c53-11e9-9608-c8901a8bfe90.gif)

## Progressive enhancement

Carousel component scales up and down according to the provided image aspect ratio:

![responsive carousel](https://user-images.githubusercontent.com/974035/59271724-9ebf7600-8c54-11e9-9d0b-21d87df825b8.gif)

**Comment by fernandofleury** [comment]
Great idea! I'm just a little bit concerned about the developer providing aspect ratio to the component

**Comment by voronianski** [comment]
> Great idea! I'm just a little bit concerned about the developer providing aspect ratio to the component

There's a default value so it can be omitted.

**Comment by connor-baer** [comment]
@fernandofleury Can you elaborate on your concern, please? Do you have an alternative solution in mind? 

AFAIK, you don't even need to use the `<SlideImage />` component — you can put anything you like inside the `<Slide />`.

**Comment by voronianski** [comment]
> @fernandofleury Can you elaborate on your concern, please? Do you have an alternative solution in mind?
> 
> AFAIK, you don't even need to use the `<SlideImage />` component — you can put anything you like inside the `<Slide />`.

true

**Comment by cristianoliveira** [comment]
WOW. That's awesome.
![giphy](https://user-images.githubusercontent.com/3959744/59281977-75a8e080-8c68-11e9-9a3b-3e0f5930c50e.gif)

One question:

Where does the `step` come from in this code below?
```js
....
  <Controls>
    <PrevButton onClick={goBack} />
    <Status step={step} total={total}  />
    <NextButton onClick={goForward} />
  </Controls>
```

**Comment by voronianski** [comment]
> One question:
> 
> Where does the `step` come from in this code below?
> 
> ```js
> ....
>   <Controls>
>     <PrevButton onClick={goBack} />
>     <Status step={step} total={total}  />
>     <NextButton onClick={goForward} />
>   </Controls>
> ```

sorry, it was not complete example, here it is -

```js
const CustomCarousel = () => {
  const total = slides.length;
  const [step, setStep] = useState(0);
  const goBack = () => setStep(step === 0 ? total - 1 : step - 1);
  const goForward = () => setStep(step === total - 1 ? 0 : step + 1);

  return (
    <Container>
      <Slides>
        {slides.map(({ image }, index) => (
          <Slide
            key={index}
            index={index}
            step={step}
          >
            <SlideImage
              src={image.src}
              alt={image.alt}
            />
          </Slide>
        ))}
      </Slides>
      <Controls>
        <PrevButton onClick={goBack} />
        <Status step={step} total={total}  />
        <NextButton onClick={goForward} />
      </Controls>
    </Container>
  );
};
```

**Comment by fernandofleury** [comment]
@connor-baer I'm missing the point of the aspect ratio. is it to apply the padding styles to the container?

**Comment by voronianski** [comment]
> @connor-baer I'm missing the point of the aspect ratio. is it to apply the padding styles to the container?

yes, it assures that image is rendered in specified proportion so we don't need to specify the exact width/height of the image.

**Comment by fernandofleury** [comment]
@voronianski That's what I thought. (I mean, I know why we use the aspect ratio to define how the image should be displayed)

I'm just thinking out loud how often we would use it. But since this has a defaultValue we should be fine. As long as we document it properly this shouldn't be an issue.

I guess I'm a bit used to those "anything sliders", you know? Where the content isn't even necessarily images

**Comment by fernandofleury** [comment]
I was also gonna propose (sorry if I'm missing it in the proposal) to be able to hide controls as well. We might want to use it as a page Hero for instance

**Comment by connor-baer** [comment]
All subcomponents are fully composable, so it's up to you to pick the ones you want to use. If you don't want any controls, then just don't use them :)

That being said, controls are crucial for accessibility and arguably for UX. They allow people who are affected by motion sickness to pause the slider and give everyone a sense of control over the action. 

We have a design for a page hero (with controls) for the marketing website: https://www.figma.com/file/JoL3G6Ma8Sq3WdLlcyCth2tJ/Marketing-Site?node-id=19%3A289

**Comment by voronianski** [comment]
Just a small update on this. Today during the hack day I started the process of porting the `Carousel` from the website to Circuit UI. 

The first step is to get rid of [dub-step](https://github.com/infiniteluke/dub-step) dependency. This npm module handles all state and swipe functionality in the current _stateful_ version of `Carousel` component but unfortunately it is not maintained anymore and has few issues as well. 

The idea here is to create improved and simplified version of it as `StepController` component inside Circuit UI and use it instead. Example of how it can look like:

```js
import { StepController } from '@sumup/circuit-ui';

function CustomCarousel() {
  return (
    <StepController
      cycle
      autoPlay
      stepDuration={1500}
      animationDuration={300}
    >
      {({
        step,
        paused,
        getStepProps,
        getNextControlProps,
        getPreviousControlProps,
        getPauseControlProps,
        getPlayControlProps
      }) => (
        {/* render here any carousel, slider, gallery components... */}
      )}
    </StepController>
  );
}
```

The second step will be to literally move all `Carousel` components from the website into Circuit UI and adjust them a little bit to match the guidelines.

Hopefully in 1 or 2 more hack days there will be a ready PR for you to review.

**Comment by connor-baer** [comment]
Hi @voronianski :) Have you made progress on a PR? If you open a draft PR, we could give some early feedback or help out.

**Comment by voronianski** [comment]
> Hi @voronianski :) Have you made progress on a PR? If you open a draft PR, we could give some early feedback or help out.

aaand here it is - https://github.com/sumup/circuit-ui/pull/482

**Comment by connor-baer** [comment]
Resolved by #482 and released in v2.0.0-alpha.2.
